### ISSUE TITLE
Compare to Branch does not focus on textbox on initial attempt

### ISSUE THREAD

**Issue by tierninho** [desktop/desktop#5600]
When selecting `Compare to Branch` from the Branch menu, the initial focus is the History tab only. Upon selecting the `Compare to Branch` again from the menu, only then does the focus shift to the cursor inside the `Select Branch to Compare...` textbox. 

**Expected result**: when selecting `Compare to Branch`, the user is taken directly to the cursor inside the `Select Branch to Compare...` textbox, as the other option is really only highlighting the History tab.

Repro:
- make sure no Compare Branch is selected in History tab
- Go to Changes tab
- select Compare to Branch from Branch menu (History tab opens)
- reselect Compare to Branch from Branch menu (cursor in `Select Branch to Compare...` textbox)

<img width="268" alt="screen shot 2018-09-11 at 8 52 59 am" src="https://user-images.githubusercontent.com/14828183/45371745-2b42e800-b5a0-11e8-8363-79773545bab8.png">

mac, 1.4.0-beta3

Logs:
```[AppStore] initializing compare state
install.ts:35 Executing getCommits: git log HEAD --date=raw --max-count=100 --pretty=%H%x1F%s%x1F%b%x1F%an <%ae> %ad%x1F%cn <%ce> %cd%x1F%P%x1F%(trailers:unfold,only) -z --no-show-signature --no-color --
install.ts:35 Executing getCommits: git log origin/hello..hello1 --date=raw --max-count=100 --pretty=%H%x1F%s%x1F%b%x1F%an <%ae> %ad%x1F%cn <%ce> %cd%x1F%P%x1F%(trailers:unfold,only) -z --no-show-signature --no-color --
install.ts:35 Executing getConfigValueInPath: git config -z trailer.separators
install.ts:35 Executing getConfigValueInPath: git config -z trailer.separators
install.ts:35 [AheadBehindUpdater] - found 0 comparisons to perform
install.ts:35 [AppStore] initializing compare state
codemirror.js:1125 [Violation] Added non-passive event listener to a scroll-blocking 'mousewheel' event. Consider marking event handler as 'passive' to make the page more responsive. See https://www.chromestatus.com/feature/5745543795965952
ss @ codemirror.js:1125
e @ simplescrollbars.js:52
n @ simplescrollbars.js:98
t.scrollbarModel.simple @ simplescrollbars.js:147
ho @ codemirror.js:3691
na @ codemirror.js:7783
na @ codemirror.js:7767
componentDidMount @ code-mirror-host.tsx:67
commitLifeCycles @ react-dom.production.min.js:159
L @ react-dom.production.min.js:185
A @ react-dom.production.min.js:182
n @ react-dom.production.min.js:181
D @ react-dom.production.min.js:181
b @ react-dom.production.min.js:180
h @ react-dom.production.min.js:179
enqueueSetState @ react-dom.production.min.js:110
a.setState @ react.production.min.js:12
e.appStore.onDidUpdate @ app.tsx:175
e.exports.e.simpleDispatch @ emitter.js:26
e.exports.e.emit @ emitter.js:144
emitUpdate @ base-store.ts:34
emitUpdateNow @ app-store.ts:424
hidden.windowState.emitQueued.window.requestAnimationFrame @ app-store.ts:416
codemirror.js:1125 [Violation] Added non-passive event listener to a scroll-blocking 'mousewheel' event. Consider marking event handler as 'passive' to make the page more responsive. See https://www.chromestatus.com/feature/5745543795965952
ss @ codemirror.js:1125
e @ simplescrollbars.js:52
n @ simplescrollbars.js:100
t.scrollbarModel.simple @ simplescrollbars.js:147
ho @ codemirror.js:3691
na @ codemirror.js:7783
na @ codemirror.js:7767
componentDidMount @ code-mirror-host.tsx:67
commitLifeCycles @ react-dom.production.min.js:159
L @ react-dom.production.min.js:185
A @ react-dom.production.min.js:182
n @ react-dom.production.min.js:181
D @ react-dom.production.min.js:181
b @ react-dom.production.min.js:180
h @ react-dom.production.min.js:179
enqueueSetState @ react-dom.production.min.js:110
a.setState @ react.production.min.js:12
e.appStore.onDidUpdate @ app.tsx:175
e.exports.e.simpleDispatch @ emitter.js:26
e.exports.e.emit @ emitter.js:144
emitUpdate @ base-store.ts:34
emitUpdateNow @ app-store.ts:424
hidden.windowState.emitQueued.window.requestAnimationFrame @ app-store.ts:416
codemirror.js:1125 [Violation] Added non-passive event listener to a scroll-blocking 'touchstart' event. Consider marking event handler as 'passive' to make the page more responsive. See https://www.chromestatus.com/feature/5745543795965952
ss @ codemirror.js:1125
oa @ codemirror.js:7860
na @ codemirror.js:7807
na @ codemirror.js:7767
componentDidMount @ code-mirror-host.tsx:67
commitLifeCycles @ react-dom.production.min.js:159
L @ react-dom.production.min.js:185
A @ react-dom.production.min.js:182
n @ react-dom.production.min.js:181
D @ react-dom.production.min.js:181
b @ react-dom.production.min.js:180
h @ react-dom.production.min.js:179
enqueueSetState @ react-dom.production.min.js:110
a.setState @ react.production.min.js:12
e.appStore.onDidUpdate @ app.tsx:175
e.exports.e.simpleDispatch @ emitter.js:26
e.exports.e.emit @ emitter.js:144
emitUpdate @ base-store.ts:34
emitUpdateNow @ app-store.ts:424
hidden.windowState.emitQueued.window.requestAnimationFrame @ app-store.ts:416
codemirror.js:1125 [Violation] Added non-passive event listener to a scroll-blocking 'touchmove' event. Consider marking event handler as 'passive' to make the page more responsive. See https://www.chromestatus.com/feature/5745543795965952
ss @ codemirror.js:1125
oa @ codemirror.js:7890
na @ codemirror.js:7807
na @ codemirror.js:7767
componentDidMount @ code-mirror-host.tsx:67
commitLifeCycles @ react-dom.production.min.js:159
L @ react-dom.production.min.js:185
A @ react-dom.production.min.js:182
n @ react-dom.production.min.js:181
D @ react-dom.production.min.js:181
b @ react-dom.production.min.js:180
h @ react-dom.production.min.js:179
enqueueSetState @ react-dom.production.min.js:110
a.setState @ react.production.min.js:12
e.appStore.onDidUpdate @ app.tsx:175
e.exports.e.simpleDispatch @ emitter.js:26
e.exports.e.emit @ emitter.js:144
emitUpdate @ base-store.ts:34
emitUpdateNow @ app-store.ts:424
hidden.windowState.emitQueued.window.requestAnimationFrame @ app-store.ts:416
codemirror.js:1125 [Violation] Added non-passive event listener to a scroll-blocking 'mousewheel' event. Consider marking event handler as 'passive' to make the page more responsive. See https://www.chromestatus.com/feature/5745543795965952
ss @ codemirror.js:1125
oa @ codemirror.js:7923
na @ codemirror.js:7807
na @ codemirror.js:7767
componentDidMount @ code-mirror-host.tsx:67
commitLifeCycles @ react-dom.production.min.js:159
L @ react-dom.production.min.js:185
A @ react-dom.production.min.js:182
n @ react-dom.production.min.js:181
D @ react-dom.production.min.js:181
b @ react-dom.production.min.js:180
h @ react-dom.production.min.js:179
enqueueSetState @ react-dom.production.min.js:110
a.setState @ react.production.min.js:12
e.appStore.onDidUpdate @ app.tsx:175
e.exports.e.simpleDispatch @ emitter.js:26
e.exports.e.emit @ emitter.js:144
emitUpdate @ base-store.ts:34
emitUpdateNow @ app-store.ts:424
hidden.windowState.emitQueued.window.requestAnimationFrame @ app-store.ts:416
install.ts:35 Executing getPartialBlobContents: git show fcd6b1be8cef32b51535ced051798294c62d9bf1:.gitignore
install.ts:35 Executing getCommits: git log HEAD --date=raw --max-count=100 --pretty=%H%x1F%s%x1F%b%x1F%an <%ae> %ad%x1F%cn <%ce> %cd%x1F%P%x1F%(trailers:unfold,only) -z --no-show-signature --no-color --
install.ts:35 Executing getConfigValueInPath: git config -z trailer.separators
install.ts:35 Executing getCommits: git log a9041631f61f678b7a9b1546f859828114ee6610^ --date=raw --max-count=100 --pretty=%H%x1F%s%x1F%b%x1F%an <%ae> %ad%x1F%cn <%ce> %cd%x1F%P%x1F%(trailers:unfold,only) -z --no-show-signature --no-color --
[AppStore] initializing compare state
install.ts:35 Executing getCommits: git log HEAD --date=raw --max-count=100 --pretty=%H%x1F%s%x1F%b%x1F%an <%ae> %ad%x1F%cn <%ce> %cd%x1F%P%x1F%(trailers:unfold,only) -z --no-show-signature --no-color --
install.ts:35 Executing getCommits: git log origin/hello..hello1 --date=raw --max-count=100 --pretty=%H%x1F%s%x1F%b%x1F%an <%ae> %ad%x1F%cn <%ce> %cd%x1F%P%x1F%(trailers:unfold,only) -z --no-show-signature --no-color --
install.ts:35 Executing getConfigValueInPath: git config -z trailer.separators
install.ts:35 Executing getConfigValueInPath: git config -z trailer.separators
install.ts:35 [AheadBehindUpdater] - found 0 comparisons to perform
install.ts:35 [AheadBehindUpdater] - found 0 comparisons to perform
install.ts:35 Executing getCommits: git log a9041631f61f678b7a9b1546f859828114ee6610^ --date=raw --max-count=100 --pretty=%H%x1F%s%x1F%b%x1F%an <%ae> %ad%x1F%cn <%ce> %cd%x1F%P%x1F%(trailers:unfold,only) -z --no-show-signature --no-color --```

**Comment by iAmWillShepherd** [comment]
Looks like this may be caused by the logic we use to focus on the text input.

https://github.com/desktop/desktop/blob/2e69e4033e3a9e5aaaf30dcb49353428c2b18c82/app/src/ui/history/compare.tsx#L95-L109

**Comment by iAmWillShepherd** [comment]
Determined the issue is not caused by the logic we use to focus on the text box. It's likely something to do with how we update state as I noticed the text input is focused before quickly loosing focus due to a re-render.

https://github.com/desktop/desktop/blob/e0d2b04b1270ccbc0ef6589fc9f572cf9dca5bd9/app/src/ui/history/compare.tsx#L90-L92

I saw that we were using an old deprecated React lifecycle method, so I went ahead and  updated it to `componentDidUpdate`, but that did not solve the underlying issue. What's happening is something is causing the component tree to re-render in what seems like an infinite loop. I've verified this behavior is caused further up the component tree, so I'll be focusing my efforts on `App.tsx` and its associated store.

Screenshot of my console after logging every call to `App::render`:

![screen shot 2018-09-19 at 2 56 32 pm](https://user-images.githubusercontent.com/1715082/45777717-3b9c3800-bc1c-11e8-9a30-8587a5f7aefc.png)
